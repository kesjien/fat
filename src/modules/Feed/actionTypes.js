export const GET_TRANSACTION_DATA = 'GET_DATA'
