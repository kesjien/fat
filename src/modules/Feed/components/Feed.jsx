import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Panel, Row, Input } from 'muicss/react'
// important here because of the bootstraping and speed, muicss input has bigger priority than
// rule: avoid important if possible
// inline styled
// in real world use another or custom implementation of inputs
const StyledRow = styled(Row)`
  margin: 1em !important; 
`

const StyledPanel = styled(Panel)`
  margin: 1em;
`
export class Feed extends React.Component {
  // define data model for better validation
  static propTypes = {
    data: PropTypes.object, // eslint-disable-line
    changeField: PropTypes.func.isRequired
  }
  static defaultProps = {
    data: {}
  }
  constructor (props) {
    super(props)
    // example of the statuses handling, should be moved to constants and
    // implemented as a list of the status
    this.cancelStatus = 'PRE_AUTHORIZED_CANCELLED'
  }

  render () {
    const { changeField, data } = this.props

    const feedComponent = []
    if (data) {
      Object.keys(data).forEach((key) => {
        if (key === 'fromDate' || key === 'toDate') {
          data[key] = data[key].slice(0, 10)
        } else if (key === 'status') { // handle status here
          // handle and parse data in the separate function, move to separate function
          if (data[key] === this.cancelStatus) {
            data[key] = 'Cancelled'
          }
        } else if (data[key] === null) {
          data[key] = ''
        }
        feedComponent.push(
          <StyledRow key={key}>
            <Input label={key} name={key} value={data[key]} onChange={changeField} />
          </StyledRow>
        )
      })
    }
    return (
      <StyledPanel>
        {feedComponent}
      </StyledPanel>
    )
  }
}

