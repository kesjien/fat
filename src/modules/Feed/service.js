import { http } from '../../services/http'
import { mappers } from '../../service-mapper'

function getFeedData (number) {
  return http.get(`/transaction/${number}`).then(dataFeed => mappers.feedMapper.mapFeedData(dataFeed.data))
}

function setFeedData(data) {
  // here should be http.put(url) to the backend,
  // since i dont have api documentation i'll omit update now
  alert('not implemented due to the lack of the API documentation, check README please')
  return true
}

export const feedSrv = {
  getFeedData,
  setFeedData
}
