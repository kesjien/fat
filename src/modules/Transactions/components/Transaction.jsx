import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Col, Row } from 'muicss/react'

const StyledRow = styled(Row)`
  border: 1px dotted #cccccc5e;
  min-height: 40px;
  text-align: center;
  margin-left: 0px !important;
  margin-right: 0px !important;
  cursor: pointer;
  &:hover {
    background-color: #f0f5fa;
  }
`

const StyledDiv = styled.div`
  border-right: 1px solid #cccccc42;
`

const StyledCol = styled(Col)`
  padding-left: 0px !important;
  padding-right: 0px !important;
`

const Transaction = (props) => {
  const cols = []
  Object.keys(props.data).forEach((key, i) => {
    cols.push(<StyledCol key={`${i}_key`} md="2"><StyledDiv className="mui--appbar-height mui--appbar-line-height">{props.data[key]}</StyledDiv></StyledCol>)
  })

  return (
    <StyledRow>
      {cols}
    </StyledRow>
  )
}

Transaction.propTypes = {
  data: PropTypes.oneOfType([
    PropTypes.array, // eslint-disable-line
    PropTypes.object // eslint-disable-line
  ]).isRequired
}

export default Transaction
