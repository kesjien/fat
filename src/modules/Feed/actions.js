import { GET_TRANSACTION_DATA } from './actionTypes'
import { feedSrv } from './service'

export const getFeedDataHandler = data => ({
  type: GET_TRANSACTION_DATA,
  payload: data
})

export function getFeedData (pageNumber) {
  return (dispatch) => {
    feedSrv.getFeedData(pageNumber)
      .then(response => dispatch(getFeedDataHandler(response)))
  }
}


export function setFeedData (data) {
  feedSrv.setFeedData(data)
}
