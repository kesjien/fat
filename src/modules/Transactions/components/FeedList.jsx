import React from 'react'
import styled from 'styled-components'
import { Link } from 'react-router-dom'
import { Panel } from 'muicss/react'
import PropTypes from 'prop-types'
import Transaction from './Transaction'

const PanelComponent = styled.div`
  margin: 1em;
  padding: 3em;
`
const StyledLink = styled(Link)`
  color: #000;
`
const Header = styled.div`
  background-color: #e6f4ff;
`

export class FeedList extends React.PureComponent {
  static propTypes = {
    // define model schema for validation transactions. temporary eslint omit line
    transactions: PropTypes.array.isRequired, // eslint-disable-line
  }
  render () {
    const transactions = []
    let newData = {}
    this.props.transactions.forEach((item) => {
      newData = {
        borrowerId: item.borrowerId,
        creditUsed: item.creditUsed,
        currency: item.currency,
        fromDate: item.fromDate.slice(0, 10),
        lenderId: item.lenderId,
        price: item.price
      }
      transactions.push(
        <div key={item.id}>
          <StyledLink to={`/${item.id}`}>
            <Transaction data={newData} />
          </StyledLink>
        </div>
      )
    })
    return (
      <Panel>
        <PanelComponent>
          <Header>
            <Transaction data={Object.keys(newData)} />
          </Header>
          {transactions}
        </PanelComponent>
      </Panel>
    )
  }
}
