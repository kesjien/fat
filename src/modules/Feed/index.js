import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Container, Button } from 'muicss/react'
import { connect } from 'react-redux'
import { getFeedData, setFeedData } from './actions'
import { Feed } from './components/Feed'

export * from './reducers'

const CustomH1 = styled.h1`
  font-size: 28px;
`

const mapStateToProps = (state) => {
  const { feed } = state.feedReducer
  return { feed }
}

const mapDispatchToProps = (dispatch) => {
  return {
    getFeedData: transactionNumber => dispatch(getFeedData(transactionNumber))
  }
}

class FeedItem extends React.Component {
  static propTypes = {
    feed: PropTypes.object, // eslint-disable-line
    transactionId: PropTypes.number,
    match: PropTypes.object.isRequired, // eslint-disable-line
    getFeedData: PropTypes.func.isRequired
  }
  static defaultProps = {
    feed: {},
    transactionId: null
  }
  constructor (props) {
    super(props)
    this.changeField = this.changeField.bind(this)
    this.setData = this.setData.bind(this)
    this.state = {
      data: {}
    }
  }
  componentDidMount () {
    const { transactionId } = this.props.match.params
    this.props.getFeedData(transactionId)
  }
  componentWillReceiveProps (props) {
    this.setState({ data: props.feed })
  }
  setData () {
    setFeedData(this.state.data)
  }
  changeField (e) {
    const { data } = this.state
    const { name, value } = e.target
    data[name] = value
    this.setState({
      data
    })
  }
  render () {
    const { data } = this.state
    let feedComponent = null
    if (Object.keys(data).length > 1) {
      feedComponent = (
        <Container fluid>
          <CustomH1>Feed data</CustomH1>
          <Feed data={data} changeField={this.changeField} />
          <Button onClick={this.setData} color="primary">Save</Button>
        </Container>
      )
    }
    return (
      <div>
        {feedComponent}
      </div>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(FeedItem)
