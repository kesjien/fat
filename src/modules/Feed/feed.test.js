import React from 'react'
import renderer from 'react-test-renderer'
import { Feed } from './components/Feed'

it('renders correctly', () => {
  const tree = renderer
    .create(<Feed data={{}} changeField={() => console.log('mocked')} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
