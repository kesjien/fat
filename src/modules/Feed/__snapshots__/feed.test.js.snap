// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="mui-panel sc-bwzfXH bGDCNZ"
/>
`;
